closure_captures.md

```diff
@@ -0,0 +1,319 @@
+ 5. Analyses the closure c2 using ExprUseVisitor and generate `upvar_capture_map`.
+ 6. c2 intravisit finishes. Done visiting the body of c1.
+ 7. Analyses the closure c2 using ExprUseVisitor and generate `upvar_capture_map`.
+ 8. Encounter the closure expression for c2. Adjust capture infromation for `a` based on `c2`
+ 8. c1 intravisit finishes
+ 9. foo intravisit finishes
+
+ More nicely (here `capture_kinds` refer to `upvar_capture_map` for a particular closure):
+```
+| visit body of foo
+| | visit body of c1
+| | | visit closure c2
+| | | | visit body of c2 (does nothing)
+| | | | analyze body of c2, producing capture_kinds[c2]
+| | analyze body of c1, reading capture_kinds[[c2], producing capture_kinds[c1]
+```
+
+#### [`Place`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/mem_categorization/struct.Place.html)
+
+Generated in [`librustc_typeck`](https://github.com/rust-lang/rust/blob/9bdd2db3a60176012f4dc240eea02d615cc60061/src/librustc_typeck/mem_categorization.rs).
+
+Several methods exist that can analyse different kinds of expression and pattern to express how memory is being accessed.
+
+### Use cases
+
+Legend:
+
+- `U`: `upvars_mentioned`
+- `C`: `closure_captures`
+- `B`: `upvar_capture_map`
+- `U->C`: Currently uses `upvars_mentioned`, but needs to start using `closure_captures`. This is most likely because we have done type checking either entierly or at least for the related scope.
+- `var_hir_id`: `HirId` of root variable of a capture
+- `CaptureKind`: If we have `var_hir_id` and access to any of the tables (i.e. access to `closure_def_id`), we can access the `CaptureKind` for the variable from `upvar_capture_map` using `UpvarId(closure_def_id, var_hir_id)`.
+- `AccessSpan`: Span in `upvars_mentioned` and points to an access of the captured variable within the closure. We can assume that we have `var_hir_id` to access this. Since we have `var_hir_id`, we also have access to definition span using using the [hir API](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.span).
+- `Field`: When we build the MIR, we desugar the closure as a structure where captures are represented by fields within the structure. We refer to these using indecies. `Field(var_hir_id)` expresses we need the `var_hir_id` of the ith capture, provided some `i`.
+- `Index`: As explained above when we build the MIR, we start representing the closure as a structure where some of the fields represent the captured variables. To build this desugared structure, we need to map `var_hir_id` to an index. For `closure_captures[c4]`, we have `hir_id_2` at index 0, and `hir_id_1` at index 1. (Order in which the captures apprear within the closure)
+
+
+| Module                 | U/C/B | Purpose                                    |  Uses    |
+|------------------------|-------|--------------------------------------------|----------|
+| [typeck/closure.rs]    | U | Initialize Substs. **TODO: #4** | `var_hir_id`
```

Comment:
This is where we make guess about types the fields that would be within the ClosureStruct when it's desugared. But based on comments around it and our initial, I think we should discuss this use case specifically.